data_managers/data_manager_fetch_genome_dbkeys_all_fasta/data_manager/data_manager_fetch_genome_all_fasta_dbkeys.py

```diff
@@ -175,155 +200,163 @@ def _sort_fasta_custom( fasta_filename, params ):
     else:
         _write_sorted_fasta( sorted_names, fasta_offsets, fasta_filename, unsorted_filename )
 
-def download_from_ucsc( data_manager_dict, params, target_directory, dbkey, dbkey_name, sequence_id, sequence_name ):
+
+def _download_file(start, fh):
+    tmp = tempfile.NamedTemporaryFile()
+    tmp.write(start)
+    tmp.write(fh.read())
+    tmp.flush()
+    tmp.seek(0)
+    return tmp
+
+
+def get_stream_reader(fh, tmp_dir):
+    """
+    Check if file is compressed and return correct stream reader.
+    If file has to be downloaded, do it now.
+    """
+    magic_dict = {
+        b"\x1f\x8b\x08": _get_stream_readers_for_gzip,
+        b"\x42\x5a\x68": _get_stream_readers_for_bz2,
+        b"\x50\x4b\x03\x04": _get_stream_readers_for_zip,
+    }
+    start_of_file = fh.read(CHUNK_SIZE)
+    try:
+        fh.seek(0)
+    except UnsupportedOperation:  # This is if fh has been created by urlopen
+        fh = _download_file(start_of_file, fh)
+    for k,v in magic_dict.items():
+        if start_of_file.startswith(k):
+            return v(fh, tmp_dir)
+    try:  # Check if file is tar file
+        if tarfile.open(fileobj=StringIO(start_of_file)):
+            return _get_stream_readers_for_tar(fh, tmp_dir)
+    except tarfile.ReadError:
+        pass
+    return fh
+
+
+def _get_ucsc_download_address(params, dbkey):
+    """
+    Check if we can find the correct file for the supplied dbkey on UCSC's FTP server
+    """
     UCSC_FTP_SERVER = 'hgdownload.cse.ucsc.edu'
     UCSC_DOWNLOAD_PATH = '/goldenPath/%s/bigZips/'
-    COMPRESSED_EXTENSIONS = [ ( '.tar.gz', _get_stream_readers_for_tar ), ( '.tar.bz2', _get_stream_readers_for_tar ), ( '.zip', _get_stream_readers_for_zip ), ( '.fa.gz', _get_stream_readers_for_gzip ), ( '.fa.bz2', _get_stream_readers_for_bz2 ) ]
-    
+    COMPRESSED_EXTENSIONS = ['.tar.gz', '.tar.bz2', '.zip', '.fa.gz', '.fa.bz2']
+
     email = params['param_dict']['__user_email__']
     if not email:
         email = 'anonymous@example.com'
 
     ucsc_dbkey = params['param_dict']['reference_source']['requested_dbkey'] or dbkey
-    UCSC_CHROM_FA_FILENAMES = [ '%s.chromFa' % ucsc_dbkey, 'chromFa', ucsc_dbkey ]
-    
-    ftp = FTP( UCSC_FTP_SERVER )
-    ftp.login( 'anonymous', email )
-    
+    UCSC_CHROM_FA_FILENAMES = ['%s.chromFa' % ucsc_dbkey, 'chromFa', ucsc_dbkey]
+
+    ftp = FTP(UCSC_FTP_SERVER)
+    ftp.login('anonymous', email)
+
     ucsc_path = UCSC_DOWNLOAD_PATH % ucsc_dbkey
-    path_contents = _get_files_in_ftp_path( ftp, ucsc_path )
-    
-    ucsc_file_name = None
-    get_stream_reader = None
-    ext = None
-    ucsc_chrom_fa_filename = None
+    path_contents = _get_files_in_ftp_path(ftp, ucsc_path)
+    ftp.close()
```

Comment:
It seems from Python doc and Google search, that `ftp.quit()` should be the proper way to close the connection, but not really sure.